src/main/java/com/phamhieu/bookapi/domain/user/UserService.java

```diff
@@ -0,0 +1,98 @@
+package com.phamhieu.bookapi.domain.user;
+
+import com.phamhieu.bookapi.persistence.user.UserStore;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+import javax.xml.bind.DatatypeConverter;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+import static com.phamhieu.bookapi.domain.user.UserError.*;
+import static com.phamhieu.bookapi.domain.user.UserValidation.*;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+@Service
+@RequiredArgsConstructor
+public class UserService {
+
+    private final UserStore userStore;
+
+    @Value("${security.encode}")
+    private String appSecurityCode;
+
+    public List<User> findAll() {
+        return userStore.findAll();
+    }
+
+    public User findById(final UUID userId) {
+        return userStore.findById(userId)
+                .orElseThrow(supplyUserNotFoundById(userId));
+    }
+
+    public List<User> find(final String userName) {
+        return userStore.find(userName);
+    }
+
+    public User create(final User user) throws NoSuchAlgorithmException {
+        validateUserInfoCreate(user);
+        verifyUsernameIfAvailable(user.getUsername());
+
+        final User tempUser = User.builder()
+                .id(user.getId())
+                .username(user.getUsername())
+                .password(hashPassword(user.getPassword()))
+                .firstName(user.getFirstName())
+                .lastName(user.getLastName())
+                .enabled(user.isEnabled())
+                .avatar(user.getAvatar())
+                .roleId(user.getRoleId())
+                .build();
+        return userStore.addUser(tempUser);
+    }
+
+    public User update(final UUID userId, final User user) throws NoSuchAlgorithmException {
+        validateUserInfoUpdate(user);
+        final User tempUser = findById(userId);
+        verifyUsernameIfAvailableUpdate(tempUser.getUsername(), user.getUsername());
```

Comment:
Tên hàm này hơi ugly

```suggestion
           if (!tempUser.getUsername().equals(user.getUsername())) {
            verifyUsernameIfAvailable(updateName);
            tempUser.setUsername(user.getUsername());
        }
```

Do username thường không phân biệt hoa thường nên phải dùng `equalsIgnoreCase` từ `StringUtils` để so sánh